src/wp-includes/class-wp-user-query.php

```diff
@@ -804,28 +804,41 @@ public function query() {
 				{$this->query_orderby}
 				{$this->query_limit}
 			";
-
-			if ( is_array( $qv['fields'] ) ) {
-				$this->results = $wpdb->get_results( $this->request );
+			$cache_key   = $this->generate_cache_key( $qv, $this->request );
+			$cache_value = wp_cache_get( $cache_key, 'users' );
+			if ( false !== $cache_value ) {
+				$this->results     = $cache_value['user_data'];
+				$this->total_users = $cache_value['total_users'];
 			} else {
-				$this->results = $wpdb->get_col( $this->request );
-			}
 
-			if ( isset( $qv['count_total'] ) && $qv['count_total'] ) {
-				/**
-				 * Filters SELECT FOUND_ROWS() query for the current WP_User_Query instance.
-				 *
-				 * @since 3.2.0
-				 * @since 5.1.0 Added the `$this` parameter.
-				 *
-				 * @global wpdb $wpdb WordPress database abstraction object.
-				 *
-				 * @param string        $sql   The SELECT FOUND_ROWS() query for the current WP_User_Query.
-				 * @param WP_User_Query $query The current WP_User_Query instance.
-				 */
-				$found_users_query = apply_filters( 'found_users_query', 'SELECT FOUND_ROWS()', $this );
-
-				$this->total_users = (int) $wpdb->get_var( $found_users_query );
+				if ( is_array( $qv['fields'] ) ) {
+					$this->results = $wpdb->get_results( $this->request );
+				} else {
+					$this->results = $wpdb->get_col( $this->request );
+				}
+
+				if ( isset( $qv['count_total'] ) && $qv['count_total'] ) {
+					/**
+					 * Filters SELECT FOUND_ROWS() query for the current WP_User_Query instance.
+					 *
+					 * @param string $sql The SELECT FOUND_ROWS() query for the current WP_User_Query.
+					 * @param WP_User_Query $query The current WP_User_Query instance.
+					 *
+					 * @global wpdb $wpdb WordPress database abstraction object.
+					 *
+					 * @since 3.2.0
+					 * @since 5.1.0 Added the `$this` parameter.
+					 *
+					 */
+					$found_users_query = apply_filters( 'found_users_query', 'SELECT FOUND_ROWS()', $this );
+
+					$this->total_users = (int) $wpdb->get_var( $found_users_query );
+				}
+				$cache_value = array(
+					'user_data'   => $this->results,
+					'total_users' => $this->total_users,
+				);
```

Comment:
I'm concerned about the storing of `user_data` here, which could store full `wp_user` DB records, significantly increasing the data stored. Especially if some code is querying for a large number of users.

`WP_Query` caching has the benefit that only post IDs are stored, thanks to the `split_the_query` functionality.

To me, it seems like this might benefit from splitting into querying for IDs and caching IDs only in a similar way, but would significantly increase the complexity of this specific PR.

Reply:
I do not believe this is the case. See https://github.com/WordPress/wordpress-develop/commit/2db7debe8a2f0b2ea264d1765c549bbf5b44a052

For all / all with meta, we only get the ids from the database for users. This means, for most cases, an array of ids will be stored and not the object. Users can filter and do weird, but this is much much less likely.

Reply:
To avoid blowing up the cache, it might be worth checking the size of `$this->query_fields[]` and turning off caching by default it it's above `n`.

Whether n is 1, 2, or 3 is a judgement call.

Reply:
> I do not believe this is the case. See https://github.com/WordPress/wordpress-develop/commit/2db7debe8a2f0b2ea264d1765c549bbf5b44a052

I'm not sure how that resolves this though? Although the main query route through it may only query by `ID` fields, it's not uncommon for there to be others.

eg, `wp_list_users()` queries by `ids` but `wp_dropdown_users()` queries by `ID, user_login` or `ID, user_login, or `ID, user_login, {$show}`.

This PR results in all of those queries using their own cache keys and values, even if what they're querying by is the same.
For an initial caching fix, this is acceptable, _but long-term, it might be more reasonable to split it into not storing  anything by the ids in the cache at all_ and then pulling the reset of the field data that's needed out from the `users` cache. Much easier now that there's cache_get_multiple to bulk-warm the users caches.

Reply:
@dd32 I believe what you mentioned is handled here - https://github.com/WordPress/wordpress-develop/commit/2db7debe8a2f0b2ea264d1765c549bbf5b44a052. 

Worth noting too `wp_dropdown_users` is used in the classic editor but no other place I could find. The User REST API and user list screen, but get fields all.